File: backend/app/api/proxy_nodes.py
"""
代理节点管理 API 路由层

路由前缀：/api/proxy-nodes
"""
import logging
from datetime import date
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, Response, UploadFile
from fastapi import status as http_status
from sqlalchemy.orm import Session
from pydantic import BaseModel

from app.core.database import get_db
from app.schemas.proxy_node import (
    ProxyNodeBatchTestResult,
    ProxyNodeCreate,
    ProxyNodeFilter,
    ProxyNodeImportResult,
    ProxyNodeResponse,
    ProxyNodeStats,
    ProxyNodeTestResult,
    ProxyNodeUpdate,
)
from app.services import proxy_node_export_service, proxy_node_import_service
from app.services import proxy_node_service, proxy_node_test_service

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/proxy-nodes", tags=["Proxy Nodes"])


# ---------------------------------------------------------------------------
# 请求体模型（仅路由层使用）
# ---------------------------------------------------------------------------

class BatchDeleteBody(BaseModel):
    node_ids: List[int]


class BatchStatusBody(BaseModel):
    node_ids: List[int]
    status: str


class BatchTestBody(BaseModel):
    node_ids: List[int]


# ---------------------------------------------------------------------------
# GET /  — 查询节点列表（分页 + 筛选）
# ---------------------------------------------------------------------------

@router.get("", response_model=dict)
def list_nodes(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=500),
    status: Optional[List[str]] = Query(None),
    protocol: Optional[List[str]] = Query(None),
    purchase_channel: Optional[str] = Query(None),
    sale_customer: Optional[str] = Query(None),
    expire_date_from: Optional[date] = Query(None),
    expire_date_to: Optional[date] = Query(None),
    db: Session = Depends(get_db),
):
    """查询节点列表，支持分页和多条件筛选。"""
    try:
        f = ProxyNodeFilter(
            status=status,
            protocol=protocol,
            purchase_channel=purchase_channel,
            sale_customer=sale_customer,
            expire_date_from=expire_date_from,
            expire_date_to=expire_date_to,
        )
        nodes, total = proxy_node_service.get_nodes(db, filter=f, skip=skip, limit=limit)
        items = [ProxyNodeResponse.model_validate(n) for n in nodes]
        return {"items": items, "total": total}
    except Exception as e:
        logger.error(f"list_nodes failed: {e}")
        raise HTTPException(
            status_code=http_status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Internal server error",
        )


# ---------------------------------------------------------------------------
# POST /  — 创建节点，返回 201
# ---------------------------------------------------------------------------

@router.post("", response_model=ProxyNodeResponse, status_code=http_status.HTTP_201_CREATED)
def create_node(data: ProxyNodeCreate, db: Session = Depends(get_db)):
    """创建单个代理节点。"""
    try:
        node = proxy_node_service.create_node(db, data)
        return node
    except Exception as e:
        logger.error(f"create_node failed: {e}")
        raise HTTPException(
            status_code=http_status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Internal server error",
        )


# ---------------------------------------------------------------------------
# GET /stats  — 获取统计数据（必须在 /{node_id} 之前注册）
# ---------------------------------------------------------------------------

@router.get("/stats", response_model=ProxyNodeStats)
def get_stats(
    expire_date_from: Optional[date] = Query(None),
    expire_date_to: Optional[date] = Query(None),
    db: Session = Depends(get_db),
):
    """获取节点统计数据，支持按到期日期范围筛选。"""
    try:
        f = ProxyNodeFilter(
            expire_date_from=expire_date_from,
            expire_date_to=expire_date_to,
        )
        return proxy_node_service.get_stats(db, filter=f)
    except Exception as e:
        logger.error(f"get_stats failed: {e}")
        raise HTTPException(
            status_code=http_status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Internal server error",
        )


# ---------------------------------------------------------------------------
# GET /import/template  — 下载导入模板 CSV
# ---------------------------------------------------------------------------

@router.get("/import/template")
def download_import_template():
    """下载节点导入模板 CSV 文件。"""
    try:
        content = proxy_node_import_service.generate_template_csv()
        return Response(
            content=content,
            media_type="text/csv; charset=utf-8",
            headers={
                "Content-Disposition": 'attachment; filename="proxy_nodes_template.csv"'
            },
        )
    except Exception as e:
        logger.error(f"download_import_template failed: {e}")
        raise HTTPException(
            status_code=http_status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Internal server error",
        )


# ---------------------------------------------------------------------------
# POST /import  — 批量导入（UploadFile）
# ---------------------------------------------------------------------------

@router.post("/import", response_model=ProxyNodeImportResult)
async def import_nodes(file: UploadFile, db: Session = Depends(get_db)):
    """批量导入节点，支持 CSV 和 Excel (.xlsx) 格式。"""
    filename = file.filename or ""
    ext = filename.rsplit(".", 1)[-1].lower() if "." in filename else ""

    if ext not in ("csv", "xlsx"):
        raise HTTPException(
            status_code=http_status.HTTP_400_BAD_REQUEST,
            detail="Unsupported file format. Use .csv or .xlsx",
        )

    try:
        content = await file.read()
    except Exception as e:
        logger.error(f"import_nodes: failed to read file: {e}")
        raise HTTPException(
            status_code=http_status.HTTP_400_BAD_REQUEST,
            detail=f"Failed to parse file: {e}",
        )

    try:
        if ext == "csv":
            result = proxy_node_import_service.import_from_csv(db, content)
        else:
            result = proxy_node_import_service.import_from_excel(db, content)
        return result
    except Exception as e:
        logger.error(f"import_nodes: parse error: {e}")
        raise HTTPException(
            status_code=http_status.HTTP_400_BAD_REQUEST,
            detail=f"Failed to parse file: {e}",
        )


# ---------------------------------------------------------------------------
# GET /export  — 导出节点数据（CSV/Excel）
# ---------------------------------------------------------------------------

@router.get("/export")
def export_nodes(
    format: str = Query("csv", pattern="^(csv|xlsx)$"),
    status: Optional[List[str]] = Query(None),
    protocol: Optional[List[str]] = Query(None),
    purchase_channel: Optional[str] = Query(None),
    sale_customer: Optional[str] = Query(None),
    expire_date_from: Optional[date] = Query(None),
    expire_date_to: Optional[date] = Query(None),
    db: Session = Depends(get_db),
):
    """导出节点数据，支持 CSV 和 Excel 格式，支持筛选条件。"""
    try:
        f = ProxyNodeFilter(
            status=status,
            protocol=protocol,
            purchase_channel=purchase_channel,
            sale_customer=sale_customer,
            expire_date_from=expire_date_from,
            expire_date_to=expire_date_to,
        )
        # 不分页，导出全部匹配节点
        nodes, _ = proxy_node_service.get_nodes(db, filter=f, skip=0, limit=100000)

        if format == "xlsx":
            file_content = proxy_node_export_service.export_to_excel(nodes)
            media_type = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
            filename = "proxy_nodes.xlsx"
        else:
            file_content = proxy_node_export_service.export_to_csv(nodes)
            media_type = "text/csv; charset=utf-8"
            filename = "proxy_nodes.csv"

        return Response(
            content=file_content,
            media_type=media_type,
            headers={
                "Content-Disposition": f'attachment; filename="{filename}"'
            },
        )
    except Exception as e:
        logger.error(f"export_nodes failed: {e}")
        raise HTTPException(
            status_code=http_status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Internal server error",
        )


# ---------------------------------------------------------------------------
# DELETE /batch  — 批量删除
# ---------------------------------------------------------------------------

@router.delete("/batch")
def batch_delete_nodes(body: BatchDeleteBody, db: Session = Depends(get_db)):
    """批量删除节点。"""
    try:
        deleted = proxy_node_service.batch_delete_nodes(db, body.node_ids)
        return {"deleted": deleted}
    except Exception as e:
        logger.error(f"batch_delete_nodes failed: {e}")
        raise HTTPException(
            status_code=http_status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Internal server error",
        )


# ---------------------------------------------------------------------------
# PATCH /batch/status  — 批量修改状态
# ---------------------------------------------------------------------------

@router.patch("/batch/status")
def batch_update_status(body: BatchStatusBody, db: Session = Depends(get_db)):
    """批量修改节点状态。"""
    try:
        updated = proxy_node_service.batch_update_status(db, body.node_ids, body.status)
        return {"updated": updated}
    except Exception as e:
        logger.error(f"batch_update_status failed: {e}")
        raise HTTPException(
            status_code=http_status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Internal server error",
        )


# ---------------------------------------------------------------------------
# POST /batch/test  — 批量测试节点连通性（必须在 /{node_id} 之前注册）
# ---------------------------------------------------------------------------

@router.post("/batch/test", response_model=ProxyNodeBatchTestResult)
async def batch_test_nodes(body: BatchTestBody, db: Session = Depends(get_db)):
    """批量测试节点连通性，最大并发 10。"""
    try:
        result = await proxy_node_test_service.batch_test_nodes(db, body.node_ids)
        return result
    except Exception as e:
        logger.error(f"batch_test_nodes failed: {e}")
        raise HTTPException(
            status_code=http_status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Internal server error",
        )


# ---------------------------------------------------------------------------
# GET /{node_id}  — 查询单个节点
# ---------------------------------------------------------------------------

@router.get("/{node_id}", response_model=ProxyNodeResponse)
def get_node(node_id: int, db: Session = Depends(get_db)):
    """按 ID 查询单个节点，不存在返回 404。"""
    node = proxy_node_service.get_node(db, node_id)
    if not node:
        raise HTTPException(
            status_code=http_status.HTTP_404_NOT_FOUND,
            detail=f"Node {node_id} not found",
        )
    return node


# ---------------------------------------------------------------------------
# PATCH /{node_id}  — 部分更新节点
# ---------------------------------------------------------------------------

@router.patch("/{node_id}", response_model=ProxyNodeResponse)
def update_node(node_id: int, data: ProxyNodeUpdate, db: Session = Depends(get_db)):
    """部分更新节点，不存在返回 404。"""
    node = proxy_node_service.update_node(db, node_id, data)
    if not node:
        raise HTTPException(
            status_code=http_status.HTTP_404_NOT_FOUND,
            detail=f"Node {node_id} not found",
        )
    return node


# ---------------------------------------------------------------------------
# DELETE /{node_id}  — 删除节点，成功返回 204
# ---------------------------------------------------------------------------

@router.delete("/{node_id}", status_code=http_status.HTTP_204_NO_CONTENT)
def delete_node(node_id: int, db: Session = Depends(get_db)):
    """删除节点，成功返回 204，不存在返回 404。"""
    success = proxy_node_service.delete_node(db, node_id)
    if not success:
        raise HTTPException(
            status_code=http_status.HTTP_404_NOT_FOUND,
            detail=f"Node {node_id} not found",
        )
    return Response(status_code=http_status.HTTP_204_NO_CONTENT)


# ---------------------------------------------------------------------------
# POST /{node_id}/test  — 测试单个节点连通性
# ---------------------------------------------------------------------------

@router.post("/{node_id}/test", response_model=ProxyNodeTestResult)
async def test_node(node_id: int, db: Session = Depends(get_db)):
    """测试单个节点连通性，不存在返回 404。"""
    result = await proxy_node_test_service.test_node(db, node_id)
    if result is None:
        raise HTTPException(
            status_code=http_status.HTTP_404_NOT_FOUND,
            detail=f"Node {node_id} not found",
        )
    return result

File: frontend/src/api/proxy_nodes.js
/**
 * Proxy Node API module
 */
import request from './request'

/**
 * 序列化数组参数为多个同名查询参数
 * axios 默认会将数组序列化为 key[]=val 格式，
 * 后端 FastAPI 期望的是多个同名参数 key=val1&key=val2
 */
function buildArrayParams(params) {
  if (!params) return params
  const result = {}
  for (const [key, value] of Object.entries(params)) {
    if (value !== undefined && value !== null && value !== '') {
      result[key] = value
    }
  }
  return result
}

/**
 * 获取节点列表（带分页和筛选）
 * @param {Object} params - skip, limit, status[], usage[], protocol[], purchase_channel, sale_customer, expire_date_from, expire_date_to
 */
export function getProxyNodes(params) {
  return request({
    url: '/proxy-nodes',
    method: 'get',
    params: buildArrayParams(params),
    // 使用 paramsSerializer 确保数组参数序列化为多个同名参数
    paramsSerializer: {
      indexes: null // 禁用索引，生成 key=val1&key=val2 格式
    }
  })
}

/**
 * 获取单个节点详情
 * @param {number|string} id - 节点 ID
 */
export function getProxyNode(id) {
  return request({
    url: `/proxy-nodes/${id}`,
    method: 'get'
  })
}

/**
 * 创建节点
 * @param {Object} data - 节点数据，ip 和 port 为必填
 */
export function createProxyNode(data) {
  return request({
    url: '/proxy-nodes',
    method: 'post',
    data
  })
}

/**
 * 更新节点（部分更新）
 * @param {number|string} id - 节点 ID
 * @param {Object} data - 要更新的字段
 */
export function updateProxyNode(id, data) {
  return request({
    url: `/proxy-nodes/${id}`,
    method: 'patch',
    data
  })
}

/**
 * 删除节点
 * @param {number|string} id - 节点 ID
 */
export function deleteProxyNode(id) {
  return request({
    url: `/proxy-nodes/${id}`,
    method: 'delete'
  })
}

/**
 * 批量删除节点
 * @param {Array<number|string>} nodeIds - 节点 ID 列表
 */
export function batchDeleteProxyNodes(nodeIds) {
  return request({
    url: '/proxy-nodes/batch',
    method: 'delete',
    data: { node_ids: nodeIds }
  })
}

/**
 * 批量修改节点状态
 * @param {Array<number|string>} nodeIds - 节点 ID 列表
 * @param {string} status - 目标状态（active/expired/sold/disabled）
 */
export function batchUpdateStatus(nodeIds, status) {
  return request({
    url: '/proxy-nodes/batch/status',
    method: 'patch',
    data: { node_ids: nodeIds, status }
  })
}

/**
 * 测试单个节点连通性
 * @param {number|string} id - 节点 ID
 */
export function testProxyNode(id) {
  return request({
    url: `/proxy-nodes/${id}/test`,
    method: 'post'
  })
}

/**
 * 批量测试节点连通性
 * @param {Array<number|string>} nodeIds - 节点 ID 列表
 */
export function batchTestProxyNodes(nodeIds) {
  return request({
    url: '/proxy-nodes/batch/test',
    method: 'post',
    data: { node_ids: nodeIds }
  })
}

/**
 * 导入节点（CSV 或 Excel 文件）
 * @param {File} file - 要上传的文件对象
 */
export function importProxyNodes(file) {
  const formData = new FormData()
  formData.append('file', file)
  return request({
    url: '/proxy-nodes/import',
    method: 'post',
    data: formData,
    headers: {
      'Content-Type': 'multipart/form-data'
    }
  })
}

/**
 * 下载导入模板 CSV
 * 通过 request 获取 blob 后创建隐藏 <a> 标签触发浏览器下载，
 * 确保请求携带 Authorization header
 */
export function downloadImportTemplate() {
  return request({
    url: '/proxy-nodes/import/template',
    method: 'get',
    responseType: 'blob'
  }).then((blob) => {
    const objectUrl = URL.createObjectURL(blob)
    const link = document.createElement('a')
    link.href = objectUrl
    link.download = 'proxy_nodes_template.csv'
    document.body.appendChild(link)
    link.click()
    document.body.removeChild(link)
    URL.revokeObjectURL(objectUrl)
  })
}

/**
 * 导出节点数据（触发文件下载）
 * @param {Object} params - 筛选参数（与 getProxyNodes 相同）
 * @param {string} format - 导出格式：'csv' 或 'xlsx'
 */
export function exportProxyNodes(params, format) {
  return request({
    url: '/proxy-nodes/export',
    method: 'get',
    params: { ...buildArrayParams(params), format },
    paramsSerializer: {
      indexes: null
    },
    responseType: 'blob'
  }).then((blob) => {
    const ext = format === 'xlsx' ? 'xlsx' : 'csv'
    const filename = `proxy_nodes_export.${ext}`
    const objectUrl = URL.createObjectURL(blob)
    const link = document.createElement('a')
    link.href = objectUrl
    link.download = filename
    document.body.appendChild(link)
    link.click()
    document.body.removeChild(link)
    URL.revokeObjectURL(objectUrl)
  })
}

/**
 * 获取节点统计数据
 * @param {Object} [params] - 可选筛选参数，包含 expire_date_from, expire_date_to
 */
export function getProxyNodeStats(params) {
  return request({
    url: '/proxy-nodes/stats',
    method: 'get',
    params: buildArrayParams(params)
  })
}
